src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -517,42 +533,39 @@ private static String toLowerCase(String s) {
         return new String(chars);
     }
 
-    /**
-     * Optionally decodes a String, if it's encoded. If String is not encoded,
-     * method will return the original input value.
-     *
-     * @param input the value String to decode
-     * @return a decoded String
-     */
-    private String percentDecode(final String input) {
-        if (input == null) {
-            return null;
-        }
-        final String decoded = uriDecode(input);
-        if (!decoded.equals(input)) {
-            return decoded;
+    public static String uriDecode(final String source) {
+        if (source == null || source.isEmpty()) {
+            return source;
         }
-        return input;
-    }
 
-    public static String uriDecode(String source) {
-        if (source == null) {
+        int percent = source.indexOf('%');
+
+        if (percent == -1) {
             return source;
         }
-        int length = source.length();
-        StringBuilder builder = new StringBuilder();
-        for (int i = 0; i < length; i++) {
-            if (source.charAt(i) == '%') {
-                String str = source.substring(i + 1, i + 3);
-                char c = (char) Integer.parseInt(str, 16);
-                builder.append(c);
-                i += 2;
-            }
-            else {
-                builder.append(source.charAt(i));
+
+        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
+        int length = bytes.length;
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream(length);
+        buffer.write(bytes, 0, percent);
+
+        for (int i = percent; i < length; i++) {
+            int b = bytes[i];
+
+            if (b == '%') {
+                    if (i + 2 >= length) {
+                        return null;
+                    }
+
+                    int b1 = Character.digit(bytes[++i], 16);
+                    int b2 = Character.digit(bytes[++i], 16);
+                    buffer.write((char) ((b1 << 4) + b2));
+            } else {
+                buffer.write(b);
             }
         }
-        return builder.toString();
+
+        return new String(buffer.toByteArray(), StandardCharsets.UTF_8);
```

Comment:
```suggestion
        return new String(bytes, 0, writePos, StandardCharsets.UTF_8);
```

Reply:
I've changed the code to use `ByteBuffer`.

Reply:
I don't understand why this should move to `ByteBuffer`. Doesn't that decrease the performance of the algorithm in comparison to https://github.com/package-url/packageurl-java/pull/160#discussion_r1988932864 - with the above suggestion as the return? With ByteBuffer we have to loop over the entire array of bytes to count the number of `%` and then make a complete copy one byte at a time. If we kept closer to your original implementation with my small suggestion we skip having to copy the first portion of the byte array and then are just shifting the values down in the existing array; never having to allocate a new buffer or loop over the bytes to count the number of `%` characters. Maybe I'm missing something.